### PR TITLE
move Subscription functions

### PR DESCRIPTION
**Pull request by avkos**
## Description

[4830](https://app.zenhub.com/workspaces/web3js-553a980ab285bae538012a40/issues/chainsafe/web3.js/4830)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [x] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [x] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

**Comment by avkos** [comment]
change name of subscriptions related to this doc https://geth.ethereum.org/docs/rpc/pubsub :
pendingTransactions => newPendingTransactions
newBlockHeaders => newHeads

File: packages/web3-eth/src/web3_eth.ts [hunk 1]
```diff
@@ -17,6 +17,13 @@ import * as rpcMethodsWrappers from './rpc_method_wrappers';
 import { Transaction, TransactionCall } from './types';
 import { Web3EthExecutionAPI } from './web3_eth_execution_api';
 
+enum SubscriptionNames {
+	logs = 'logs',
+	newPendingTransactions = 'newPendingTransactions',
```
Comment (jdevcs) [inline packages/web3-eth/src/web3_eth.ts]:
```suggestion
	pendingTransactions = 'pendingTransactions',
```

as per 1.x

File: packages/web3-eth/src/web3_eth.ts [hunk 2]
```diff
@@ -17,6 +17,13 @@ import * as rpcMethodsWrappers from './rpc_method_wrappers';
...
+	newHeads = 'newHeads',
```
Comment (jdevcs) [inline packages/web3-eth/src/web3_eth.ts]:
```suggestion
	newBlockHeaders = 'newBlockHeaders',
```

**Review by jdevcs — CHANGES_REQUESTED** [review]

File: packages/web3-eth/src/web3_eth.ts (hunk 1 above)
Comment (avkos) [reply packages/web3-eth/src/web3_eth.ts]: I've described this:
change name of subscriptions related to this doc https://geth.ethereum.org/docs/rpc/pubsub :
pendingTransactions => newPendingTransactions
newBlockHeaders => newHeads

And, also in 1.x version, we have comments to rename these things

Comment (avkos) [reply packages/web3-eth/src/web3_eth.ts]: we can support both options. what do you think @jdevcs?

File: packages/web3-core/src/web3_context.ts [hunk 3]
```diff
@@ -98,9 +104,23 @@ export class Web3Context<
...
 	}
 
+	private static getDefaultRegisteredSubscriptions(): any {
+		return {
+			logs: LogsSubscription,
+			newPendingTransactions: NewPendingTransactionsSubscription,
+			newHeads: NewHeadsSubscription,
+			syncing: SyncingSubscription,
+		};
+	}
```
Comment (nazarhussain) [inline packages/web3-core/src/web3_context.ts]: This could be false positive, as the user can pass `registeredSubscriptions` to the context, even user can pass `subscriptionManager`. So the values returned by this function will be false positive.

File: packages/web3-core/src/web3_context.ts [hunk 4]
```diff
@@ -98,9 +104,23 @@ export class Web3Context<
 				this.requestManager,
 				registeredSubscriptions,
 			);
+		} else {
+			this._subscriptionManager = new Web3SubscriptionManager(
+				this.requestManager,
+				Web3Context.getDefaultRegisteredSubscriptions(),
+			);
```
Comment (nazarhussain) [inline packages/web3-core/src/web3_context.ts]: If you want to set as default behavior, use it as default value of `registeredSubscriptions`

Comment (nazarhussain) [reply packages/web3-core/src/web3_context.ts]: I would not recommend it though, every context should be responsible to register their own set of subscriptions.

**Review by nazarhussain — CHANGES_REQUESTED** [review]

File: packages/web3-core/src/web3_context.ts (hunk 4 above)
Comment (avkos) [reply packages/web3-core/src/web3_context.ts]: Thanks. You are right. I will add a setter for `registeredSubscriptions` and pass them from web3_eth package

File: packages/web3-eth/src/web3_eth.ts (hunk 1 above)
Comment (avkos) [reply packages/web3-eth/src/web3_eth.ts]: I've added both options to support

File: packages/web3-eth/src/web3_eth.ts (hunk 2 above)
Comment (avkos) [reply packages/web3-eth/src/web3_eth.ts]: I've added both options to support

File: packages/web3-core/src/web3_context.ts [hunk 5]
```diff
@@ -135,6 +135,15 @@ export class Web3Context<
...
 	}
 
+	protected setRegisteredSubscriptions(registeredSubscriptions: RegisteredSubs): void {
+		if (!this.subscriptionManager) {
+			this._subscriptionManager = new Web3SubscriptionManager(
+				this.requestManager,
+				registeredSubscriptions,
+			);
+		}
+	}
```
Comment (nazarhussain) [inline packages/web3-core/src/web3_context.ts]: The type `RegisteredSubs` is generic and needs to be passed in the class definition. So does not seems correct to  specify type (which is subscription classes) at definition and actual values later on. The type `RegisteredSubs` is generic and needs to be passed in the class definition. So does not seems correct to  specify type (which is subscription classes) at definition and actual values later on.

File: packages/web3-core/src/web3_subscription_manager.ts [hunk 6]
```diff
@@ -78,16 +84,17 @@ export class Web3SubscriptionManager<
...
 		}
-
+		const { id } = sub;
 		await sub.unsubscribe();
-		this._subscriptions.delete(sub.id);
+		this._subscriptions.delete(id);
 	}
 
-	public async unsubscribe() {
+	public async unsubscribe(condition?: ShouldUnsubscribeCondition) {
```
Comment (nazarhussain) [inline packages/web3-core/src/web3_subscription_manager.ts:92]: The function `web3.eth.subscribe` returns a subscription object. If user want to clear that one particular subscription, he should use `unsubscribe` on that object. 

https://web3js.readthedocs.io/en/v1.5.0/web3-eth-subscribe.html#subscribe

On the Subscription Manager level, I would suggest to not add such behavior, it will be duplicate implementation and difficult to maintain in future.

File: packages/web3-core/src/web3_subscriptions.ts [hunk 7]
```diff
@@ -73,28 +72,42 @@ export abstract class Web3Subscription<
...
-	public async unsubscribe() {
-		if (!this.id) {
-			return;
-		}
-
-		await this._requestManager.send({
-			method: 'eth_unsubscribe',
-			params: [this.id] as Web3APIParams<API, 'eth_unsubscribe'>,
-		});
+	public async unsubscribe(cb?: Callback) {
```
Comment (nazarhussain) [inline packages/web3-core/src/web3_subscriptions.ts]: We decided to remove the callback support from 4.x, so why it was not added earlier. Please revert these changes.

File: packages/web3-eth/src/web3_eth.ts [hunk 8]
```diff
@@ -15,8 +20,45 @@ import * as rpcMethods from './rpc_methods';
...
+	public constructor(providerOrContext: SupportedProviders<any> | Web3ContextInitOptions) {
+		super(providerOrContext);
+		this.setRegisteredSubscriptions({
+			logs: LogsSubscription,
+			newPendingTransactions: NewPendingTransactionsSubscription,
+			newHeads: NewHeadsSubscription,
+			syncing: SyncingSubscription,
+			pendingTransactions: NewPendingTransactionsSubscription, // the same as newPendingTransactions. just for support API like in version 1.x
+			newBlockHeaders: NewHeadsSubscription, // the same as newHeads. just for support API like in version 1.x
+		});
```
Comment (nazarhussain) [inline packages/web3-eth/src/web3_eth.ts]: I am not sure if this works else where outside constructor considering the type safety. So having this method does not seems right. Would suggest to pass these to the `super` parameters.

File: packages/web3-eth/src/web3_eth.ts [hunk 9]
```diff
@@ -249,4 +291,69 @@ export class Web3Eth extends Web3Context<Web3EthExecutionAPI> {
...
+		if (typeof cb === 'function') {
+			return { options: args as LogArguments, callBack: cb };
+		}
+		if (typeof args === 'function') {
+			return { options: {} as LogArguments, callBack: args };
+		}
+		return {
+			options: args as LogArguments,
+		};
+	}
```
Comment (nazarhussain) [inline packages/web3-eth/src/web3_eth.ts]: Move this as utility function outside class.

File: packages/web3-eth/src/web3_eth.ts [hunk 10]
```diff
@@ -249,4 +291,69 @@ export class Web3Eth extends Web3Context<Web3EthExecutionAPI> {
...
+
+	public async subscribe(
+		name: keyof typeof SubscriptionNames,
+		args?: LogArguments | Callback | undefined,
+		cb?: Callback,
+	): Promise<SomeSubscription> {
```
Comment (nazarhussain) [inline packages/web3-eth/src/web3_eth.ts]: This will not return `SomeSubscription`, instead exactly the one matching with the `name`.

File: packages/web3-eth/src/web3_eth.ts [hunk 11]
```diff
@@ -249,4 +291,69 @@ export class Web3Eth extends Web3Context<Web3EthExecutionAPI> {
...
+			return { options: {} as LogArguments, callBack: args };
+		}
+		return {
+			options: args as LogArguments,
+		};
+	}
+
+	public async subscribe(
+		name: keyof typeof SubscriptionNames,
+		args?: LogArguments | Callback | undefined,
```
Comment (nazarhussain) [inline packages/web3-eth/src/web3_eth.ts]: We decided to remove the support for callbacks from `4.x`, so why its' not implemented in subscription manager and subscription class. Please remove this support from here as well.

Comment (nazarhussain) [inline packages/web3-eth/src/web3_eth.ts]: Don't need to specify new type `LogArguments` here, it's not typesafe.  The arguments must match the exact type of the constructor parameters of subscription class matching to the `name`. 

Please see the reference implementation of `subscribe` for the `subscription manager`. In here you just need to proxy that function, no need of any implementation.

File: packages/web3-eth/src/web3_eth.ts [hunk 12]
```diff
@@ -249,4 +291,69 @@ export class Web3Eth extends Web3Context<Web3EthExecutionAPI> {
...
+		}
+
+		return subscription as SomeSubscription;
+	}
+
+	private static shouldClearSubscription({ sub }: { sub: unknown }): boolean {
+		return !(sub instanceof SyncingSubscription);
+	}
+
+	public clearSubscriptions(notClearSyncing = false): Promise<void[]> | undefined {
```
Comment (nazarhussain) [inline packages/web3-eth/src/web3_eth.ts:321]: I don't see any reference of this parameter in the docs

https://web3js.readthedocs.io/en/v1.5.0/web3-eth-subscribe.html#clearsubscriptions

**Review by nazarhussain — CHANGES_REQUESTED** [review]

File: packages/web3-core/src/web3_subscriptions.ts (hunk 7 above)
Comment (avkos) [reply packages/web3-core/src/web3_subscriptions.ts]: I don't know this. And make all things by documentation. But I absolutely agree with you :)

Comment (avkos) [reply packages/web3-core/src/web3_subscriptions.ts:75]: Did we also decide to remove callback from `subscribe` method? @nazarhussain

File: packages/web3-eth/src/web3_eth.ts (hunk 12 above)
Comment (avkos) [reply packages/web3-eth/src/web3_eth.ts:321]: please read here https://web3js.readthedocs.io/en/v1.7.1/web3-eth-subscribe.html#clearsubscriptions

File: packages/web3-eth/src/web3_eth.ts [hunk 13]
```diff
@@ -249,4 +283,45 @@ export class Web3Eth extends Web3Context<Web3EthExecutionAPI> {
...
+					.then(logs => {
+						for (const log of logs) {
+							subscription._processSubscriptionResult(log);
+						}
+					})
+					.catch(e => {
+						subscription._processSubscriptionError(e as Error);
+					});
+			});
+		}
```
Comment (nazarhussain) [inline packages/web3-eth/src/web3_eth.ts:313]: Can we do this inside `LogsSubscription` implementation? If yes may be a separate issue will be good to refactor it.

**Review by nazarhussain — APPROVED** [review]

File: packages/web3-eth/src/web3_eth.ts (hunk 13 above)
Comment (avkos) [reply packages/web3-eth/src/web3_eth.ts:313]: yes we can move into LogsSubscription, but in this case we need to have duplicate of implementation `getPastLogs` method inside LogsSubscription

File: tools/isomorphic-ws.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
+var ws = null;
+
+if (typeof WebSocket !== 'undefined') {
```
Comment (jdevcs) [inline tools/isomorphic-ws.js:3]: Isomorphic lib should automatically switch to ws in case of node.js and to global.WebSocket for browsers.

**Review by luu-alex — APPROVED** [review]

File: tools/isomorphic-ws.js (hunk 14 above)
Comment (jdevcs) [inline tools/isomorphic-ws.js:3]: Isomorphic lib should automatically switch to ws in case of node.js and to global.WebSocket for browsers. Why we need to add this logic explicitly here.

Comment (avkos) [reply tools/isomorphic-ws.js:3]: it is just a copy of isomorphic lib. but I've just added this line `module.exports.WebSocket = ws;`
because isomorphic for NodeJS and for webrowser has different external interface
for browser 
`import WebSocket from 'isomorphic'`
for nodejs
`import {WebSocket} from 'isomorphic'`

File: package.json [hunk 15]
```diff
@@ -37,7 +37,8 @@
 		"ganache": "npx ganache -m 'smart guide what forget tired jungle always expire rescue boring glue champion' --account=\"0x4c3758228f536f7a210f8936182fb5b728046970b8e3215d0b5cb4c4faae8a4e,100000000000000000000\" --account=\"0x34aeb1f338c17e6b440c189655c89fcef148893a24a7f15c0cb666d9cf5eacb3,100000000000000000000\"",
 		"ganache-background": "yarn ganache > /dev/null & client=$! npx wait-port 8545",
 		"geth:build": "docker build ./.github -t gethnode -f .github/Dockerfile",
-		"geth": "docker run -d -p 8545:8545 gethnode --datadir data --networkid 15 --http.addr 0.0.0.0 --http.port 8545 --http",
+		"geth": "docker run -p 8545:8545 gethnode --datadir data --networkid 15 --http --http.addr 0.0.0.0 --http.port 8545 --ws --ws.addr 0.0.0.0 --ws.port 8545",
```
Comment (jdevcs) [inline package.json]:
```suggestion
		"geth": "docker run -p 8545:8545 -p 8546:8546 gethnode --datadir data --networkid 15 --http --http.addr 0.0.0.0 --http.port 8545 --ws --ws.addr 0.0.0.0 --ws.port 8546 --http.api personal,web3,eth,admin,debug,miner,txpool,net --ws.api personal,web3,eth,admin,debug,miner,txpool,net --nodiscover",
```

File: package.json [hunk 16]
```diff
@@ -37,7 +37,8 @@
...
+		"geth-background": "docker run -d -p 8545:8545 gethnode --datadir data --networkid 15 --http --http.addr 0.0.0.0 --http.port 8545 --ws --ws.addr 0.0.0.0 --ws.port 8545",
```
Comment (jdevcs) [inline package.json]:
```suggestion
		"geth-background": "docker run -d -p 8545:8545 -p 8546:8546 gethnode --datadir data --networkid 15 --http --http.addr 0.0.0.0 --http.port 8545 --ws --ws.addr 0.0.0.0 --ws.port 8546 --http.api personal,web3,eth,admin,debug,miner,txpool,net --ws.api personal,web3,eth,admin,debug,miner,txpool,net --nodiscover",
```

**Review by jdevcs — CHANGES_REQUESTED** [review]

**Review by jdevcs — APPROVED** [review]
